lib/ide/utils/src/handle.rs

```diff
@@ -0,0 +1,268 @@
+    /// Likely the object (if there was an object) has been already dropped.
+    HandleExpired,
+    /// RefCell was already borrowed, cannot `borrow_mut`. Should never happen.
+    AlreadyBorrowed,
+}
+
+
+
+// ================
+// == WeakHandle ==
+// ================
+
+/// A type that can provide `Weak` handle to a value under `RefCell`.
+pub trait IsWeakHandle: Clone {
+    /// Type of the data stored.
+    type Data;
+
+    /// Obtain weak handle to the data.
+    fn weak_handle(&self) -> WeakHandle<Self::Data>;
+
+    /// Obtain strong handle to the data.
+    fn upgrade(&self) -> Option<StrongHandle<Self::Data>> {
+        self.weak_handle().upgrade()
+    }
+
+    /// Returns `Future` that shall try calling `Cb` with the access to mutably
+    /// borrowed data.
+    fn with<Cb,R> (&self,cb:Cb) -> WeakWith<Self,Cb>
+        where Cb: FnOnce(&mut Self::Data) -> R {
+        WeakWith::new(self.clone(),cb)
+    }
+}
+
+/// Clones the value loaded from the handle.
+pub async fn try_load<Data: Clone>
+(handle:&impl IsWeakHandle<Data=Data>) -> Result<Data> {
+    handle.with(|data| data.clone()).await
+}
+
+/// Stores the value into the handle.
+pub async fn try_store<Data>
+(handle:&impl IsWeakHandle<Data=Data>, value:Data) -> Result<()> {
+    handle.with(|data| *data = value).await
+}
+
+impl<T> IsWeakHandle for WeakHandle<T> {
+    type Data = T;
+    fn weak_handle(&self) -> WeakHandle<Self::Data> {
+        self.clone()
+    }
+}
+
+
+
+// ==================
+// == StrongHandle ==
+// ==================
+
+/// A type that can provide a strong (`Rc`) handle to a value under `RefCell`.
+pub trait IsStrongHandle: Clone {
+    /// Type of the data stored.
+    type Data;
+
+    /// Obtain strong (`Rc`) handle to the data.
+    fn strong_handle(&self) -> StrongHandle<Self::Data>;
+
+    /// Obtain weak handle to the data.
+    fn downgrade(&self) -> WeakHandle<Self::Data> {
+        Rc::downgrade(&self.strong_handle())
+    }
+
+    /// Returns `Future` that shall call `Cb` with the access to mutably
+    /// borrowed data.
+    fn with<Cb,R> (&self, cb:Cb) -> StrongWith<Self,Cb>
+    where Cb: FnOnce(&mut Self::Data) -> R {
+        StrongWith::new(self.clone(),cb)
+    }
+}
+
+/// Clones the value loaded from the handle.
+pub async fn load<Data: Clone>(handle:&impl IsStrongHandle<Data=Data>) -> Data {
+    handle.with(|data| data.clone()).await
+}
+
+/// Stores the value into the handle.
+pub async fn store<Data>(handle:&impl IsStrongHandle<Data=Data>, value:Data) {
+    handle.with(|data| *data = value).await
+}
+
+impl<T> IsStrongHandle for StrongHandle<T> {
+    type Data = T;
+    fn strong_handle(&self) -> StrongHandle<Self::Data> {
+        self.clone()
+    }
+}
+
+
+
+// ==============
+// == WeakWith ==
+// ==============
+
+/// Future that shall call `cb` with the data under `handle`.
+#[derive(Debug)]
+pub struct WeakWith<Handle,Cb> {
+    handle : Handle,
+    cb     : Option<Cb>,
```

Comment:
This sohuld be named `callback`. Its not an internal variable inside of a function. It is even mentioned in documentation.

Reply:
Done